File: server/src/routes/bank-account-settings.js
const router = require('express').Router();
const db = require('../db');
const asyncHandler = require('../asyncHandler');
const { pickFields, requireRole } = require('../utils');

const FIELDS = ['bank_name', 'bank_bin', 'account_number', 'account_holder_name', 'min_withdrawal_balance'];

/** Chỉ giữ 1 dòng đang áp dụng — dòng mới nhất theo updated_at. */
async function currentSettings() {
  return db.queryOne('SELECT * FROM bank_account_settings ORDER BY updated_at DESC LIMIT 1');
}

// Công khai — app khách cần đọc để tự dựng URL ảnh VietQR ở màn chi tiết đơn (đơn
// bank_transfer), không riêng gì admin.
router.get('/bank-account-settings', asyncHandler(async (req, res) => {
  const row = await currentSettings();
  res.json({ ok: true, data: row });
}));

router.patch('/bank-account-settings', asyncHandler(async (req, res) => {
  requireRole(req.ctx, ['admin']);
  const existing = await currentSettings();
  const data = { ...pickFields(req.body, FIELDS), updated_at: new Date().toISOString(), updated_by: req.ctx.userId };
  const updated = existing
    ? await db.updateById('bank_account_settings', existing.id, data)
    : await db.insertRow('bank_account_settings', data);
  res.json({ ok: true, data: updated });
}));

module.exports = router;

File: hofa_customer_app/lib/models/bank_account_settings.dart
/// Thông tin tài khoản ngân hàng của sàn — dùng dựng URL ảnh VietQR ở màn chi tiết đơn cho
/// đơn thanh toán bằng chuyển khoản. Chỉ đọc (admin mới sửa được, xem hofa_admin_app).
class BankAccountSettings {
  final String? bankName;
  final String? bankBin;
  final String? accountNumber;
  final String? accountHolderName;

  BankAccountSettings({
    this.bankName,
    this.bankBin,
    this.accountNumber,
    this.accountHolderName,
  });

  bool get isConfigured =>
      (bankBin != null && bankBin!.isNotEmpty) && (accountNumber != null && accountNumber!.isNotEmpty);

  factory BankAccountSettings.fromJson(Map<String, dynamic> json) => BankAccountSettings(
        bankName: json['bank_name'] as String?,
        bankBin: json['bank_bin'] as String?,
        accountNumber: json['account_number'] as String?,
        accountHolderName: json['account_holder_name'] as String?,
      );

  factory BankAccountSettings.empty() => BankAccountSettings();
}
